hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala

```diff
@@ -218,19 +224,28 @@ class JSBuilder(using TL, State, Ctx) extends CodeBuilder:
                 .mkDocument(doc"")
             val preCtorCode = ctorAuxParams.flatMap(ps => ps).foldLeft(body(preCtor, endSemi = true)):
               case (acc, (sym, nme)) =>
-                doc"$acc # this.${sym.name} = $nme;"
-            val ctorCode = doc"$preCtorCode${body(ctor, endSemi = false)}"
+                doc"$acc # this${fieldSelect(sym.name)} = $nme;"
+            val ctorCode = doc"$preCtorCode${body(ctor, endSemi = auxParams.nonEmpty)}"
 
-            val ctorBod = if auxParams.isEmpty then
-              doc"${braced(ctorCode)}"
+            val ctorAux = if auxParams.isEmpty then
+              ctorCode
             else
               val pss = ctorAuxParams.map(_.map(_._2))
               val newCtorCode = doc"$ctorCode # return this;"
               val ctorBraced = doc"${ braced(newCtorCode) }"
               val funBod = pss.foldRight(ctorBraced):
                 case (psDoc, doc) => doc"(${psDoc.mkDocument(", ")}) => $doc"
 
-              doc"${ braced(doc" # return $funBod") }"
+              doc" # return $funBod";
+            
+            val ctorBod = if isModule then
```

Comment:
Just a thought: could a similar change be applied for objects?

Reply:
No, because the constructor of object could be used by somewhere else...
That probably means the fix for module might be incorrect too? if module constructor could be used multiple times

Reply:
and another thing: I just tested it and it also caused problem with inheritance, as you can only access `this` after the super call.

Reply:
But the
```scala
              case N =>
                braced(doc" # ${getVar(sym)} = ${getVar(isym)};$ctorCode")
```
case still made sense, right?

To me that just shows that we should always insert that (even when there is an owner) and make sure self-references refer to that variable while we're in the ctor before `super`, no?

Reply:
Oh, I am probably confusing myself here. The change (hack) was that we assign the original name to the name generated by JSCodegen.

So what I currently know is that this fix cannot be applied to object, as its constructor might be called to create new instances that is not the original instance and also this change insert it at the beginning, meaning that super is not called yet.

However, module "constructor" is just the static block currently, which is probably fine to have the assignment there? Even if it has an owner the code in the static block will only run once instead of the case for the object.

Reply:
I'm not sure whether module will have its constructor exposed and called in a similar way to how object can. If it is the case, then applying this hack will cause problem later.

Reply:
Ok @AnsonYeung, what's preventing us from just adding a similar early-initializer for `object` but just after the `super` call and using `this` instead of the module/object's name?

Reply:
The constructor might be called again, but we do not want to change the binding when a new object is constructed

Reply:
Ah, right. We could check and only assign when whether the binding is `undefined`, but that's a bit hacky. Perhaps better to leave it as is.